### DISCUSSION TITLE
not pattern does not compile with same named class

### DISCUSSION THREAD

**Discussion by AnakinRaW** [dotnet/csharplang#4237] [General] [answered]
Consider the following code: 
```cs
object o = new not();
if (o is not) { } // Compile error
if (o is not not) { } // Works fine

public class not { }
```
On VS 16.8.3 Roslyn gives me `CS8504`. 

Are there any plans on fixing this (e.g. by making `not` a keyword) or was this intentionally not considered (e.g. due to backwards compatibility)?

**Comment by 333fred** [comment, marked as answer]
This type of scenario was intentionally not (no pun intended) considered. Lowercase type names are bad style in C#, and while we make every effort to not break this type of scenario, it can happen (for example, in C# 9, putting `record foo;` in a type always means generate a record named `foo`, regardless of whether there is a type also named `record`). For your scenario, escaping with `@` works:

```cs
object o = null;
_ = o is @not;

public class not
{
}
```
